src/theory/quantifiers/sygus/enum_stream_concrete.h

```diff
@@ -0,0 +1,303 @@
+ *
+ * When a given enumerator is "variable agnostic", only values in which
+ * variables are ordered are chosen for it (see
+ * TermDbSygus::registerEnumerator). If such values are seen as "abstract", in
+ * the sense that each represent a set of values, this class can be used to
+ * stream all the concrete values that correspond to them.
+ *
+ * For example a variable agnostic enumerator that contains three variables, x1,
+ * x2, x3, in which x1 < x2 < x3, for which an "abstract" value (OP x1 x2) is
+ * derived, will lead to the stream of "concrete" values
+ *
+ * (OP x1 x2)
+ * (OP x1 x3)
+ * (OP x2 x3)
+ *
+ * (OP x2 x1)
+ * (OP x3 x1)
+ * (OP x3 x2)
+ *
+ * in which for each permutation of the variables in the abstract value
+ * ([x1, x2] and [x2, x1] in the above) we generate all the ordered combinations
+ * of the variables of the enumerator ([x1, x2], [x1, x3], and [x2, x3] in the
+ * above).
+ *
+ * Moreover the permutations are generated modulo rewriting, s.t. if two
+ * permutations are equivalent, only one is used.
+ *
+ * It should be noted that the variables of a variable agnostic enumerator are
+ * kept in independent "subclasses" (see TermDbSygus::getSubclassForVar).
+ * Therefore when streaming concrete values permutations and combinations are
+ * generated by the product of the permutations and combinations of each class.
+ */
+class EnumStreamConcrete
+{
+ public:
+  EnumStreamConcrete(QuantifiersEngine* qe, SynthConjecture* p);
+  ~EnumStreamConcrete() {}
+  /** initializes class with the given enumerator
+   *
+   * during initialization builds a map from the variables of the enumerator's
+   * type to their (type) subclasses
+   */
+  void initialize(Node e);
+  /** registers abstract value v
+   *
+   * during registration collects variables occurring in value and sets up
+   * utilities for streaming combinations and permutations for building concrete
+   * values
+   */
+  void registerAbstractValue(Node v);
+  /** generates next concrete value
+   *
+   * if no more combinations / permutations exist from the last registered
+   * value, this method returns Node::null()
+   */
+  Node getNext();
+
+ private:
+  /** reference to quantifier engine */
+  QuantifiersEngine* d_qe;
+  /** reference to the parent conjecture */
+  SynthConjecture* d_parent;
+  /** sygus term database of d_qe */
+  quantifiers::TermDbSygus* d_tds;
+  /** enumerator we are concretizing values for */
+  Node d_enum;
+  /** variables from enumerator's type */
+  std::vector<Node> d_vars;
+  /** partition of variables per subclasses */
+  std::vector<std::vector<Node>> d_var_classes;
+  /** maps variables to their respective constructors in all the enumerator
+   * sutypes */
+  std::map<Node, std::vector<Node>> d_var_cons;
+  /** list of registered abstract values */
+  std::vector<Node> d_abs_values;
+  /** combination utils of registered values */
+  std::vector<StreamCombination> d_stream_combinations;
```

Comment:
Shouldnt we only have a combination util for the most recent value? I'm a bit worried about memory consumption here.

Reply:
I was thinking that when we change the variables for the combinations we could reuse some of the information already computed for previous values.